src/main/java/uk/gov/hmcts/reform/feature/security/CustomUserRolesFilter.java

```diff
@@ -4,54 +4,64 @@
 import org.springframework.security.access.intercept.RunAsUserToken;
 import org.springframework.security.authentication.AnonymousAuthenticationToken;
 import org.springframework.security.core.Authentication;
-import org.springframework.security.core.AuthenticationException;
 import org.springframework.security.core.GrantedAuthority;
 import org.springframework.security.core.authority.SimpleGrantedAuthority;
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.security.core.userdetails.User;
 import org.springframework.security.core.userdetails.UserDetails;
-import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter;
 import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.web.filter.GenericFilterBean;
 import uk.gov.hmcts.reform.feature.model.UserRoles;
 
 import java.io.IOException;
 import java.util.List;
+import javax.servlet.FilterChain;
 import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
 import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
 
 import static org.springframework.http.HttpMethod.GET;
 
-public class CustomUserRolesFilter extends AbstractAuthenticationProcessingFilter {
+public class CustomUserRolesFilter extends GenericFilterBean {
 
-    static final String USER_ID_HEADER = "X-USER-ID";
+    public static final String USER_ID_HEADER = "X-USER-ID";
 
-    static final String USER_ROLES_HEADER = "X-USER-ROLES";
+    public static final String USER_ROLES_HEADER = "X-USER-ROLES";
+
+    private final RequestMatcher matcher;
 
     public CustomUserRolesFilter(String pattern) {
-        super(new AntPathRequestMatcher(pattern, GET.name()));
+        matcher = new AntPathRequestMatcher(pattern, GET.name());
     }
 
     @Override
-    public Authentication attemptAuthentication(
-        HttpServletRequest request,
-        HttpServletResponse response
-    ) throws AuthenticationException, IOException, ServletException {
-        String userIdHeader = request.getHeader(USER_ID_HEADER);
-        String userRolesHeader = request.getHeader(USER_ROLES_HEADER);
-        Authentication originalAuth = SecurityContextHolder.getContext().getAuthentication();
-
-        if (checkHeadersAreValid(userIdHeader, userRolesHeader)) {
-            UserRoles userRoles = new UserRoles(
-                userIdHeader,
-                userRolesHeader.split(",")
-            );
-
-            return parseUserRoles(userRoles, originalAuth);
+    public void doFilter(
```

Comment:
Instead of implementing filter could we not do something like below.
1.Disable FF4J REST API to read feature toggle values.
2.Implement custom endpoint to read feature toggle value

  ```@GetMapping(path = "/api/ff4j/check/{featureUid}")
    public ResponseEntity<Boolean> retrieveFeatureToggle(@PathVariable String featureUid, @RequestHeader("X-USER-ROLES") String roles) throws IOException {
        List<String> userRoles = Arrays.asList(roles.split(","));

        Set<String> permissions = ff4j.getFeature(featureUid).getPermissions();

        return ok(!Collections.disjoint(userRoles,permissions));
    }
}
```

Reply:
nice suggestion. but not only permissions are in play. strategies are checked too. this approach ignores them all.
not mentioning all the change in functional tests.

on the other hand it will fix problem with swagger docs. moreover it goes back to the idea of having a wrapper over the library - what we thought in the first place.

in short, it can be tricky and other teams already about to integrate with this api. not entirely sure it is safe to separate API dependency out. must be carefully weighed

Reply:
Totally agree with you @doncem